src/lib/rules/apple-touch-icons/apple-touch-icons.ts

```diff
@@ -0,0 +1,268 @@
+
+            try {
+                networkData = await context.fetchContent(appleTouchIconURL);
+            } catch (e) {
+                debug(`Failed to fetch the ${appleTouchIconHref} file`);
+                await context.report(resource, appleTouchIcon, `'${appleTouchIconHref}' file request failed`);
+
+                return;
+            }
+
+            const response = networkData.response;
+
+            if (response.statusCode !== 200) {
+                await context.report(resource, appleTouchIcon, `'${appleTouchIconHref}' could not be fetched (status code: ${response.statusCode})`);
+
+                return;
+            }
+
+            // - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+            // Note: Async version of `image-size` doesn't work if the
+            // input is a Buffer.
+            //
+            // https://github.com/image-size/image-size/tree/4c527ba608d742fbb29f6d9b3c77b831b069cbb2#asynchronous
+
+            const image = getImageData(response.body.rawContent);
+
+            // Check if the image is a PNG.
+
+            if (image.type !== 'png') {
+                await context.report(resource, appleTouchIcon, `'${appleTouchIconHref}' is not a PNG`);
+            }
+
+            // Check if the image is 180x180px.
+
+            if (image.width !== 180 || image.height !== 180) {
+                await context.report(resource, appleTouchIcon, `'${appleTouchIconHref}' is not 180x180px`);
+            }
+
+            // TODO: Check if the image has some kind of transparency.
+        };
+
+        const chooseBestIcon = (icons: Array<IAsyncHTMLElement>): IAsyncHTMLElement => {
+
+            // Site will usually have something such as:
+            //
+            // <link rel="apple-touch-icon" sizes="60x60" href="/apple-touch-icon-60x60.png">
+            // <link rel="apple-touch-icon" sizes="72x72" href="/apple-touch-icon-72x72.png">
+            // <link rel="apple-touch-icon" sizes="76x76" href="/apple-touch-icon-76x76.png">
+            // <link rel="apple-touch-icon" sizes="114x114" href="/apple-touch-icon-114x114.png">
+            // <link rel="apple-touch-icon" sizes="120x120" href="/apple-touch-icon-120x120.png">
+            // <link rel="apple-touch-icon" sizes="144x144" href="/apple-touch-icon-144x144.png">
+            // <link rel="apple-touch-icon" sizes="152x152" href="/apple-touch-icon-152x152.png">
+            // <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon-180x180.png">
+            // <link rel="apple-touch-icon" href="/apple-touch-icon-57x57.png">
+            //
+            // so what this function will try to do is select the
+            // icon that will most likely generate the fewest errors.
+
+            let bestIcon;
+
+            for (const icon of icons) {
+                const sizes = normalizeString(icon.getAttribute('sizes'));
+
+                if (!sizes) {
+                    return icon;
```

Comment:
What do you think about `continue`ing here instead of returning. It could be possible that one of the other ones has `180x180`. Also, should we warn if they don't have sizes? I'm not too familiar with what is a best practice here, but just throwing out some thoughts.

Reply:
> What do you think about continueing here instead of returning. It could be possible that one of the other ones has 180x180

That was intentional. I want the one without `sizes` to be chosen over the `180x180`. But thinking again about it, yes, maybe it will be better to choose the one with `sizes="180x180"`since in most cases, it will probably actually be `180x180px`, and that will require less work.

> Also, should we warn for those tags that do not have the sizes attribute? I'm not too familiar with what is a best practice here, but just throwing out some thoughts.

Since we are guiding the user to just one icon, using [`sizes` doesn't make much sense](https://github.com/sonarwhal/sonar/pull/488/files#diff-75ae57f8862939f10ef6a8409a4b777cR69).

Reply:
> maybe it will be better to choose the one with `sizes="180x180"`

Done.